File: src/MixNetFx/ADSqlSupportComponents/AdSQLDirectoryPicker.vb
Imports System.IO
Imports System.Data.SqlClient
Imports System.Collections.ObjectModel

Public Class ADSQLDirectoryPicker
    Inherits TreeView

    Private myConnectionString As String
    Private myDrives As Collection(Of DBDriveItem)
    Private myImageList As ImageList
    Private myExtensionFilter As String
    Private myUpdateBlocked As Boolean
    Private myUpdateOnUpdateUnblocked As Boolean

    Public Event SelectedFileNodeChanged(ByVal sender As Object, ByVal e As ADFileTreeViewEventArgs)

    Sub New()
        MyBase.New()
        myImageList = New ImageList
        myImageList.Images.Add(My.Resources.drive, Color.Magenta)
        myImageList.Images.Add(My.Resources.VSFolder_closed, Color.Magenta)
        myImageList.Images.Add(My.Resources.document, Color.Magenta)
        Me.ImageList = myImageList
        myExtensionFilter = Nothing
    End Sub

    Public Shadows Sub BeginUpdate()
        MyBase.BeginUpdate()
        myUpdateBlocked = True
    End Sub

    Public Shadows Sub EndUpdate()
        MyBase.EndUpdate()
        myUpdateBlocked = False
        If myUpdateOnUpdateUnblocked Then
            rebuildRootList()
        End If
    End Sub

    Public Property ConnectionString() As String
        Get
            Return myConnectionString
        End Get
        Set(ByVal value As String)
            myConnectionString = value
            If Not myUpdateBlocked Then
                rebuildRootList()
            Else
                myUpdateOnUpdateUnblocked = True
            End If
        End Set
    End Property

    Public Property ExtensionFilter() As String
        Get
            Return myExtensionFilter
        End Get
        Set(ByVal value As String)
            myExtensionFilter = value
            If Not myUpdateBlocked Then
                rebuildRootList()
            Else
                myUpdateOnUpdateUnblocked = True
            End If
        End Set
    End Property

    Private Function testConnection() As Boolean
        Dim locConnection As New SqlConnection(myConnectionString)
        Dim locIcon As MessageBoxIcon = MessageBoxIcon.Exclamation
        Using locConnection
            Try
                locConnection.Open()
            Catch ex As Exception
                Dim locMessage As String = "Verbindungsherstellung war nicht möglich!" & _
                    vbNewLine & vbNewLine & ex.Message & vbNewLine & vbNewLine & _
                    ex.StackTrace
                locIcon = MessageBoxIcon.Error
                MessageBox.Show(locMessage, "Verbindungstest:", MessageBoxButtons.OK, locIcon)
                Return False
            End Try
        End Using
        Return True
    End Function

    Private Sub rebuildRootList()
        'Knoten Löschen
        Me.Nodes.Clear()
        If String.IsNullOrEmpty(myConnectionString) Then
            Return
        End If
        If testConnection() Then
            'Wurzel (=Laufwerke ermitteln) und darstellen
            myDrives = ADSqlDriveFoldersAndFilesListing.GetDrivenames(myConnectionString)
            For Each locItem As DBDriveItem In myDrives
                Dim locNode As TreeNode = Me.Nodes.Add(locItem.DriveLetter, locItem.DriveLetter & ":", 0, 0)
                'Erste Ebene einlesen
                Dim locDirItems As New Collection(Of DBDirOrFileItem)
                locDirItems = ADSqlDriveFoldersAndFilesListing.GetSubfoldersAndFiles(myConnectionString, locItem.DriveLetter & ":\")
                buildSubnode(locNode, locDirItems)
            Next
        End If
    End Sub

    Protected Overrides Sub OnBeforeExpand(ByVal e As System.Windows.Forms.TreeViewCancelEventArgs)
        MyBase.OnBeforeExpand(e)
        For Each locNode As TreeNode In e.Node.Nodes
            locNode.Nodes.Clear()
            Dim locDirItems As New Collection(Of DBDirOrFileItem)
            locDirItems = ADSqlDriveFoldersAndFilesListing.GetSubfoldersAndFiles(myConnectionString, locNode.FullPath)
            buildSubnode(locNode, locDirItems)
        Next
    End Sub

    Protected Overrides Sub OnAfterCollapse(ByVal e As System.Windows.Forms.TreeViewEventArgs)
        MyBase.OnAfterCollapse(e)
        Dim locNode As TreeNode = e.Node
        locNode.Nodes.Clear()
        Dim locDirItems As New Collection(Of DBDirOrFileItem)
        locDirItems = ADSqlDriveFoldersAndFilesListing.GetSubfoldersAndFiles(myConnectionString, locNode.FullPath)
        buildSubnode(locNode, locDirItems)
    End Sub

    Private Sub buildSubnode(ByVal node As TreeNode, ByVal dirItems As Collection(Of DBDirOrFileItem))
        If dirItems IsNot Nothing Then
            For Each DirItem As DBDirOrFileItem In dirItems
                If DirItem.IsFile And (Not String.IsNullOrEmpty(ExtensionFilter)) Then
                    Dim locFileInfo As New FileInfo(DirItem.Name)
                    If locFileInfo.Extension = ExtensionFilter Or ExtensionFilter = ".*" Then
                        node.Nodes.Add(DirItem.Name, DirItem.Name, 2, 2)
                    End If
                ElseIf Not DirItem.IsFile Then
                    node.Nodes.Add(DirItem.Name, DirItem.Name, 1, 1)
                End If
            Next
        End If
    End Sub

    Protected Overrides Sub OnAfterSelect(ByVal e As System.Windows.Forms.TreeViewEventArgs)
        MyBase.OnAfterSelect(e)
        Dim locE As ADFileTreeViewEventArgs
        If e.Node IsNot Nothing Then
            locE = New ADFileTreeViewEventArgs(e.Node, e.Action, CType(e.Node.ImageIndex, ADFileItemType))
        Else
            locE = New ADFileTreeViewEventArgs(e.Node, e.Action, ADFileItemType.None)
        End If
        OnSelectedFileNodeChanged(locE)
    End Sub

    Protected Overridable Sub OnSelectedFileNodeChanged(ByVal e As ADFileTreeViewEventArgs)
        RaiseEvent SelectedFileNodeChanged(Me, e)
    End Sub
End Class

Public Class ADFileTreeViewEventArgs
    Inherits TreeViewEventArgs

    Private myFileItemType As ADFileItemType

    Sub New(ByVal node As TreeNode, ByVal action As TreeViewAction, ByVal itemType As ADFileItemType)
        MyBase.New(node, action)
        myFileItemType = itemType
    End Sub

    Public Property FileItemType() As ADFileItemType
        Get
            Return myFileItemType
        End Get
        Set(ByVal value As ADFileItemType)
            myFileItemType = value
        End Set
    End Property
End Class

Public Enum ADFileItemType
    Drive
    Folder
    File
    None
End Enum

File: src/NetFxCSharp/ADSqlSupportComponents/AdSQLDirectoryPicker.cs
using System.Collections.ObjectModel;
using System.Drawing;
using System.IO;
using System.Windows.Forms;

namespace ActiveDev.Data.SqlClient
{
    public class ADSQLDirectoryPicker : TreeView
    {
        private string _connectionString;
        private Collection<DBDriveItem> _drives;
        private readonly ImageList _imageList;
        private string _extensionFilter;
        private bool _updateBlocked;
        private bool _updateOnUpdateUnblocked;

        public event System.EventHandler<ADFileTreeViewEventArgs> SelectedFileNodeChanged;

        public ADSQLDirectoryPicker() : base()
        {
            _imageList = new ImageList();
            _imageList.Images.Add(global::ActiveDev.Data.SqlClient.My.Resources.drive, Color.Magenta);
            _imageList.Images.Add(global::ActiveDev.Data.SqlClient.My.Resources.VSFolder_closed, Color.Magenta);
            _imageList.Images.Add(global::ActiveDev.Data.SqlClient.My.Resources.document, Color.Magenta);
            this.ImageList = _imageList;
            _extensionFilter = null;
        }

        public new void BeginUpdate()
        {
            base.BeginUpdate();
            _updateBlocked = true;
        }

        public new void EndUpdate()
        {
            base.EndUpdate();
            _updateBlocked = false;
            if (_updateOnUpdateUnblocked)
                RebuildRootList();
        }

        public string ConnectionString
        {
            get => _connectionString;
            set
            {
                _connectionString = value;
                if (!_updateBlocked)
                    RebuildRootList();
                else
                    _updateOnUpdateUnblocked = true;
            }
        }

        public string ExtensionFilter
        {
            get => _extensionFilter;
            set
            {
                _extensionFilter = value;
                if (!_updateBlocked)
                    RebuildRootList();
                else
                    _updateOnUpdateUnblocked = true;
            }
        }

        private bool TestConnection()
        {
            using (var connection = new System.Data.SqlClient.SqlConnection(_connectionString))
            {
                var icon = MessageBoxIcon.Exclamation;
                try
                {
                    connection.Open();
                }
                catch (System.Exception ex)
                {
                    string msg = "Verbindungsherstellung war nicht möglich!" +
                        "\r\n\r\n" + ex.Message + "\r\n\r\n" + ex.StackTrace;
                    icon = MessageBoxIcon.Error;
                    MessageBox.Show(msg, "Verbindungstest:", MessageBoxButtons.OK, icon);
                    return false;
                }
            }
            return true;
        }

        private void RebuildRootList()
        {
            this.Nodes.Clear();
            if (string.IsNullOrEmpty(_connectionString))
                return;

            if (TestConnection())
            {
                _drives = ADSqlDriveFoldersAndFilesListing.GetDrivenames(_connectionString);
                foreach (DBDriveItem item in _drives)
                {
                    TreeNode node = this.Nodes.Add(item.DriveLetter, item.DriveLetter + ":", 0, 0);
                    var dirItems = ADSqlDriveFoldersAndFilesListing.GetSubfoldersAndFiles(_connectionString, item.DriveLetter + ":\\");
                    BuildSubnode(node, dirItems);
                }
            }
        }

        protected override void OnBeforeExpand(TreeViewCancelEventArgs e)
        {
            base.OnBeforeExpand(e);
            foreach (TreeNode node in e.Node.Nodes)
            {
                node.Nodes.Clear();
                var dirItems = ADSqlDriveFoldersAndFilesListing.GetSubfoldersAndFiles(_connectionString, node.FullPath);
                BuildSubnode(node, dirItems);
            }
        }

        protected override void OnAfterCollapse(TreeViewEventArgs e)
        {
            base.OnAfterCollapse(e);
            TreeNode node = e.Node;
            node.Nodes.Clear();
            var dirItems = ADSqlDriveFoldersAndFilesListing.GetSubfoldersAndFiles(_connectionString, node.FullPath);
            BuildSubnode(node, dirItems);
        }

        private void BuildSubnode(TreeNode node, Collection<DBDirOrFileItem> dirItems)
        {
            if (dirItems != null)
            {
                foreach (DBDirOrFileItem item in dirItems)
                {
                    if (item.IsFile && !string.IsNullOrEmpty(ExtensionFilter))
                    {
                        var fileInfo = new FileInfo(item.Name);
                        if (fileInfo.Extension == ExtensionFilter || ExtensionFilter == ".*")
                            node.Nodes.Add(item.Name, item.Name, 2, 2);
                    }
                    else if (!item.IsFile)
                    {
                        node.Nodes.Add(item.Name, item.Name, 1, 1);
                    }
                }
            }
        }

        protected override void OnAfterSelect(TreeViewEventArgs e)
        {
            base.OnAfterSelect(e);
            ADFileTreeViewEventArgs args;
            if (e.Node != null)
                args = new ADFileTreeViewEventArgs(e.Node, e.Action, (ADFileItemType)e.Node.ImageIndex);
            else
                args = new ADFileTreeViewEventArgs(e.Node, e.Action, ADFileItemType.None);
            OnSelectedFileNodeChanged(args);
        }

        protected virtual void OnSelectedFileNodeChanged(ADFileTreeViewEventArgs e)
        {
            SelectedFileNodeChanged?.Invoke(this, e);
        }
    }

    public class ADFileTreeViewEventArgs : TreeViewEventArgs
    {
        private ADFileItemType _fileItemType;

        public ADFileTreeViewEventArgs(TreeNode node, TreeViewAction action, ADFileItemType itemType)
            : base(node, action)
        {
            _fileItemType = itemType;
        }

        public ADFileItemType FileItemType
        {
            get => _fileItemType;
            set => _fileItemType = value;
        }
    }

    public enum ADFileItemType
    {
        Drive,
        Folder,
        File,
        None
    }
}
